File: java/100/LeetCode0121.java
public class LeetCode0121 {
    /**
     * 交易一次的最大利润
     * 在卖出前，应该选择之前的日子中价格最低的一点买入
     */
    public int maxProfit(int[] prices) {
        if (prices == null || prices.length == 0) {
            return 0;
        }
        int profit = 0;
        int buy = prices[0];
        for (int i = 1;i < prices.length;++i) {
            if (prices[i] > buy) {
                profit = Math.max(profit, prices[i] - buy);
            } else if (prices[i] < buy) {
                buy = prices[i];
            }
        }
        return profit;
    }
}

File: cpp/L100/LeetCode0121.cc
#include <vector>
using std::vector;

class Solution {
public:
    int maxProfit(vector<int>& prices) {
        if (prices.empty()) {
            return 0;
        }
        int profit{0}, min{prices[0]};
        for (auto it{prices.begin() + 1};it != prices.end();++it) {
            if (*it > min) {
                profit = std::max(profit, *it - min);
            } else if (*it < min) {
                min = *it;
            }
        }
        return profit;
    }
};